test/e2e/cloud-provider-azure.go

```diff
@@ -44,7 +45,7 @@ func InstallCalicoAndCloudProviderAzureHelmChart(ctx context.Context, input clus
 	specName := "cloud-provider-azure-install"
 	By("Installing the correct version of cloud-provider-azure components via helm")
 	options := &helmVals.Options{
-		Values: []string{fmt.Sprintf("infra.clusterName=%s", input.ConfigCluster.ClusterName)},
+		Values: []string{fmt.Sprintf("infra.clusterName=%s", input.ConfigCluster.ClusterName), fmt.Sprintf("cloudControllerManager.clusterCIDR=%s", strings.Join(cidrBlocks, `,`))},
```

Comment:
@lzhecheng I'd like to add a test that tests the behavior that would be broken if this wasn't there but I'm a bit confused by the fact that our existing test with the out-of-tree cloud provider (https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/test/e2e/azure_test.go#L533) doesn't fail since it exposes a service via public LB. Do you have any suggestions on what test we could add?

Reply:
Does "if this wasn't there" mean cluster CIDR is incorrect for CCM? I think CCM will crash and keep restarting in that case with logs like:
```
F1202 02:28:14.102691       1 controllermanager.go:362] error running controllers: failed to mark cidr[2001:1234:5678:9a40::/64] at idx [0] as occupied for node: zhecheng-1129-control-plane-b4x5z: cidr 2001:1234:5678:9a40::/64 is out the range of cluster cidr 2001:1234:5679:9a40::/58
```

Reply:
@lzhecheng here are the logs of a CCM pod from a test that ran on another PR (so _without_ this change): https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-azure/2890/pull-cluster-api-provider-azure-e2e-optional/1598699416011673600/artifacts/clusters/capz-e2e-oj2ty8-oot/kube-system/cloud-controller-manager-747f499866-dhcbj/cloud-controller-manager.log

``` 
I1202 15:39:50.378876       1 flags.go:64] FLAG: --cluster-cidr="10.244.0.0/16"
``` 

https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-azure/2890/pull-cluster-api-provider-azure-e2e-optional/1598699416011673600/artifacts/clusters/bootstrap/resources/capz-e2e-oj2ty8/Cluster/capz-e2e-oj2ty8-oot.yaml

``` 
spec:
  clusterNetwork:
    pods:
      cidrBlocks:
      - 192.168.0.0/16
 ``` 

Clearly the cluster-cidr (default one) passed into CCM is different from the actual cluster cidr. I'm not seeing any errors or restarts in the CCM pod and even see successful logs like:

``` 
I1202 15:40:53.974989       1 azure_routes.go:442] CreateRoute: route created. clusterName="capz-e2e-oj2ty8-oot" instance="capz-e2e-oj2ty8-oot-control-plane-nrgfm" cidr="10.244.0.0/24"
```

Why is this not failing as you described above?

Reply:
I created a cluster `zhecheng-1205` with `ci-entrypoint.sh` locally. Cluster config says pod cidr is `192.168.0.0/16`:
```
zhechengli@devbox:~/github.com/lzhecheng/cloud-provider-azure/kubetest2-aks$ kd cluster zhecheng-1205
Name:         zhecheng-1205
Namespace:    default
Labels:       cni=zhecheng-1205-calico
Annotations:  <none>
API Version:  cluster.x-k8s.io/v1beta1
Kind:         Cluster
...
Spec:
  Cluster Network:
    Pods:
      Cidr Blocks:
        192.168.0.0/16
  Control Plane Endpoint:
    Host:  zhecheng-1205-3415dece.eastus.cloudapp.azure.com
    Port:  6443
  Control Plane Ref:
    API Version:  controlplane.cluster.x-k8s.io/v1beta1
    Kind:         KubeadmControlPlane
    Name:         zhecheng-1205-control-plane
    Namespace:    default
  Infrastructure Ref:
    API Version:  infrastructure.cluster.x-k8s.io/v1beta1
    Kind:         AzureCluster
    Name:         zhecheng-1205
    Namespace:    default
Status:
  Conditions:
    Last Transition Time:  2022-12-05T02:19:40Z
    Status:                True
    Type:                  Ready
    Last Transition Time:  2022-12-05T02:19:40Z
    Status:                True
    Type:                  ControlPlaneInitialized
    Last Transition Time:  2022-12-05T02:19:40Z
    Status:                True
    Type:                  ControlPlaneReady
    Last Transition Time:  2022-12-05T02:17:12Z
    Status:                True
    Type:                  InfrastructureReady
  Control Plane Ready:     true
  Failure Domains:
    1:
      Control Plane:  true
    2:
      Control Plane:  true
    3:
      Control Plane:     true
  Infrastructure Ready:  true
  Observed Generation:   2
  Phase:                 Provisioned
Events:                  <none>
```

However, if I describe the node, the pod cidr is `10.244.0.0/24`:
```
zhechengli@devbox:~/github.com/lzhecheng/cluster-api-provider-azure$ kgno
NAME                                STATUS   ROLES           AGE     VERSION
zhecheng-1205-control-plane-tdglz   Ready    control-plane   3h36m   v1.25.3
zhecheng-1205-md-0-9zg2s            Ready    <none>          3h35m   v1.25.3
zhecheng-1205-md-0-wcbsv            Ready    <none>          3h35m   v1.25.3
zhechengli@devbox:~/github.com/lzhecheng/cluster-api-provider-azure$ kdno zhecheng-1205-control-plane-tdglzName:               zhecheng-1205-control-plane-tdglz
Roles:              control-planeLabels:             beta.kubernetes.io/arch=amd64
                    beta.kubernetes.io/instance-type=Standard_D2s_v3                    beta.kubernetes.io/os=linux
                    failure-domain.beta.kubernetes.io/region=eastus                    failure-domain.beta.kubernetes.io/zone=eastus-2
                    kubernetes.io/arch=amd64
                    kubernetes.io/hostname=zhecheng-1205-control-plane-tdglz
                    kubernetes.io/os=linux
                    node-role.kubernetes.io/control-plane=
                    node.kubernetes.io/exclude-from-external-load-balancers=
                    node.kubernetes.io/instance-type=Standard_D2s_v3
                    topology.kubernetes.io/region=eastus
                    topology.kubernetes.io/zone=eastus-2
Annotations:        cluster.x-k8s.io/cluster-name: zhecheng-1205
                    cluster.x-k8s.io/cluster-namespace: default
                    cluster.x-k8s.io/machine: zhecheng-1205-control-plane-scfbc
                    cluster.x-k8s.io/owner-kind: KubeadmControlPlane
                    cluster.x-k8s.io/owner-name: zhecheng-1205-control-plane
                    kubeadm.alpha.kubernetes.io/cri-socket: unix:///var/run/containerd/containerd.sock
                    node.alpha.kubernetes.io/ttl: 0
                    projectcalico.org/IPv4Address: 10.0.0.4/16
                    projectcalico.org/IPv4VXLANTunnelAddr: 192.168.159.0
                    volumes.kubernetes.io/controller-managed-attach-detach: true
CreationTimestamp:  Mon, 05 Dec 2022 02:19:24 +0000
Taints:             node-role.kubernetes.io/control-plane:NoSchedule
Unschedulable:      false
Lease:
  HolderIdentity:  zhecheng-1205-control-plane-tdglz
  AcquireTime:     <unset>
  RenewTime:       Mon, 05 Dec 2022 05:56:14 +0000
Conditions:
  Type                 Status  LastHeartbeatTime                 LastTransitionTime                Reason                       Message
  ----                 ------  -----------------                 ------------------                ------                       -------
  NetworkUnavailable   False   Mon, 05 Dec 2022 02:20:01 +0000   Mon, 05 Dec 2022 02:20:01 +0000   CalicoIsUp                   Calico is running on this node
  MemoryPressure       False   Mon, 05 Dec 2022 05:53:11 +0000   Mon, 05 Dec 2022 02:19:24 +0000   KubeletHasSufficientMemory   kubelet has sufficient memory available
  DiskPressure         False   Mon, 05 Dec 2022 05:53:11 +0000   Mon, 05 Dec 2022 02:19:24 +0000   KubeletHasNoDiskPressure     kubelet has no disk pressure  PIDPressure          False   Mon, 05 Dec 2022 05:53:11 +0000   Mon, 05 Dec 2022 02:19:24 +0000   KubeletHasSufficientPID      kubelet has sufficient PID available  Ready                True    Mon, 05 Dec 2022 05:53:11 +0000   Mon, 05 Dec 2022 02:19:59 +0000   KubeletReady                 kubelet is posting ready status. AppArmor enabledAddresses:
  InternalIP:  10.0.0.4  Hostname:    zhecheng-1205-control-plane-tdglz
Capacity:  cpu:                2
  ephemeral-storage:  129886128Ki  hugepages-1Gi:      0
  hugepages-2Mi:      0  memory:             8142840Ki
  pods:               110Allocatable:
  cpu:                2  ephemeral-storage:  119703055367
  hugepages-1Gi:      0  hugepages-2Mi:      0
  memory:             8040440Ki
  pods:               110
System Info:
  Machine ID:                 3d167097a8854698842f146b258aa02a
  System UUID:                45a30f33-0804-0a49-922d-3938d134f4eb
  Boot ID:                    f4d4fb70-9b5d-402e-89f4-d2501dccc6c3
  Kernel Version:             5.15.0-1021-azure
  OS Image:                   Ubuntu 20.04.5 LTS
  Operating System:           linux
  Architecture:               amd64
  Container Runtime Version:  containerd://1.6.2
  Kubelet Version:            v1.25.3
  Kube-Proxy Version:         v1.25.3
PodCIDR:                      10.244.0.0/24
PodCIDRs:                     10.244.0.0/24
ProviderID:                   azure:///subscriptions/8ecadfc9-d1a3-4ea4-b844-0d9f87e4d7c8/resourceGroups/zhecheng-1205/providers/Microsoft.Compute/virtualMachines/zhecheng-1205-control-plane-tdglz
Non-terminated Pods:          (8 in total)
  Namespace                   Name                                                         CPU Requests  CPU Limits  Memory Requests  Memory Limits  Age
  ---------                   ----                                                         ------------  ----------  ---------------  -------------  ---
  kube-system                 calico-node-bbfhv                                            250m (12%)    0 (0%)      0 (0%)           0 (0%)         3h36m
  kube-system                 cloud-controller-manager-75d946776c-hgnjg                    100m (5%)     4 (200%)    128Mi (1%)       2Gi (26%)      33m
  kube-system                 cloud-node-manager-5dmtt                                     50m (2%)      2 (100%)    50Mi (0%)        512Mi (6%)     33m
  kube-system                 etcd-zhecheng-1205-control-plane-tdglz                       100m (5%)     0 (0%)      100Mi (1%)       0 (0%)         3h36m
  kube-system                 kube-apiserver-zhecheng-1205-control-plane-tdglz             250m (12%)    0 (0%)      0 (0%)           0 (0%)         3h36m
  kube-system                 kube-controller-manager-zhecheng-1205-control-plane-tdglz    200m (10%)    0 (0%)      0 (0%)           0 (0%)         3h36m
  kube-system                 kube-proxy-5zrvw                                             0 (0%)        0 (0%)      0 (0%)           0 (0%)         3h36m
  kube-system                 kube-scheduler-zhecheng-1205-control-plane-tdglz             100m (5%)     0 (0%)      0 (0%)           0 (0%)         3h36m
Allocated resources:
  (Total limits may be over 100 percent, i.e., overcommitted.)
  Resource           Requests     Limits
  --------           --------     ------
  cpu                1050m (52%)  6 (300%)
  memory             278Mi (3%)   2560Mi (32%)
  ephemeral-storage  0 (0%)       0 (0%)
  hugepages-1Gi      0 (0%)       0 (0%)
  hugepages-2Mi      0 (0%)       0 (0%)
Events:
```

Somehow node's PodCIDR doesn't match cluster pod cidr. I think you can reproduce it.

Reply:
@lzhecheng I looked into this a bit more an it seems that's it's the CCM cluster CIDR that influences what the node pod CIDR is and not the other way around (and then the actual pods that get scheduled get IPs from the CNI pod CIDR...)

I created two clusters:

- One without explicitly setting the cluster-cidr when installing CCM: **external-cloud-provider-31602**
- One where I set the CCM cluster-cidr to match the Cluster pod CIDR (which is also the same pod CIDR calico has) **external-cloud-provider-11551**

For the  **external-cloud-provider-31602** cluster where CCM CIDR was _not_ set:

Cluster CIDR is 192.168.0.0/16
```
k get Cluster external-cloud-provider-31602 -o template --template {{.spec.clusterNetwork.pods.cidrBlocks}}
[192.168.0.0/16]
```

Calico IPPool is the same
```
 k get ippools.crd.projectcalico.org -A -o yaml | grep cidr
    cidr: 192.168.0.0/16
``` 

New pods get IPs in that range:
``` 
k get pod nginx -o wide
NAME    READY   STATUS    RESTARTS   AGE   IP               NODE                                       NOMINATED NODE   READINESS GATES
nginx   1/1     Running   0          46m   192.168.79.194   external-cloud-provider-31602-md-0-rdp8w   <none>           <none>
``` 

But, CCM cluster-cidr is the default value:
```
 k describe po cloud-controller-manager-6f65d65fc-5q2rd -n kube-system | grep cluster-cidr
      --cluster-cidr=10.244.0.0/16
 ``` 

AND the node pod CIDRs are in the range of the above value:

``` 
k get nodes -o yaml | grep podCIDR:
    podCIDR: 10.244.0.0/24
    podCIDR: 10.244.1.0/24
    podCIDR: 10.244.2.0/24
``` 

Now for the  **external-cloud-provider-11551** cluster where CCM CIDR did get configured to the value of the Cluster pod CIDR:

Cluster CIDR is 192.168.0.0/16
```
k get cluster external-cloud-provider-11551 -o template --template {{.spec.clusterNetwork.pods.cidrBlocks}}
[192.168.0.0/16]                         
```

Calico IPPool is the same
```
 k get ippools.crd.projectcalico.org -A -o yaml | grep cidr
    cidr: 192.168.0.0/16
``` 

New pods get IPs in that range:
``` 
k get pod nginx -o wide
NAME    READY   STATUS    RESTARTS   AGE   IP               NODE                                       NOMINATED NODE   READINESS GATES
nginx   1/1     Running   0          23s   192.168.210.66   external-cloud-provider-11551-md-0-xdnhm   <none>           <none>
``` 

CCM cluster-cidr is the expected value:
```
k describe po cloud-controller-manager-567c7c94dd-df5d6 -n kube-system | grep cluster-cidr
      --cluster-cidr=192.168.0.0/16
 ``` 

AND the node pod CIDRs are in the range of the above value:

``` 
k get nodes -o yaml | grep podCIDR:
    podCIDR: 192.168.0.0/24
    podCIDR: 192.168.1.0/24
    podCIDR: 192.168.2.0/24
``` 

So what this tells me is that it's actually CCM that sets the value of the node podCIDR or that value is at least derived from the CCM cluster-cidr value since that's the only thing that changed between the two cases. In both cases pods scheduled on the nodes get a pod IP in the CNI pod CIDR pool. What I still don't understand is why things don't seem to be broken in the case where there is a mismatch between node podCIDR and CNI/Cluster podCIDR, that's where my Kubernetes networking knowledge falls short. 

@aramase does any of this surprise you or am I misunderstanding something?

Reply:
Now I got it. Yes, CCM sets the Node PodCIDR and before CCM starts, Nodes have no PodCIDR.
I created a cluster with `TEST_CCM=true` and remove those lines in script to helm install CCM. Then I checked the created cluster and found that Nodes have no PodCIDR. Then I helm install CCM and they appeared. I checked CCM log and found that it set those values.
```
I1206 02:14:05.166592       1 range_allocator.go:277] Putting node zhecheng-1206-control-plane-ffl4b with CIDR [10.244.0.0/24] into the work queue
I1206 02:14:05.166663       1 range_allocator.go:277] Putting node zhecheng-1206-md-0-wpwdv with CIDR [10.244.1.0/24] into the work queue
I1206 02:14:05.166702       1 range_allocator.go:277] Putting node zhecheng-1206-md-0-k7w89 with CIDR [10.244.2.0/24] into the work queu
```
Code is here:
https://github.com/kubernetes-sigs/cloud-provider-azure/blob/2eef98fca741842fe2831454afa7b6458ecf3873/pkg/nodeipam/ipam/range_allocator.go#L257-L278
The logic is that CCM sets Nodes' PodCIDR if there's none.